### PR TITLE
Rewrite

### PR DESCRIPTION
**Pull request by Adman**
to @mrshu

cc @Krista @pepol @JakubNVK @tWido @Foxiy

### CODE REVIEW

File: karma.py [hunk 1]
```diff
@@ -0,0 +1,84 @@
...
+
+        if user not in karmas:
+            karmas[user] = 0
+
+        self['karma'] = karmas
+        return "{0}'s karma level is: {1}".format(user, karmas[user])
+
+    @botcmd
+    def top_karma(self, msg, args):
+        """Get 5 people with most karma points."""
```
Comment (mrshu) [inline karma.py]: @Adman this does not seem to be what the function does now, in particular the `5 people` part.

File: karma.py [hunk 2]
```diff
@@ -0,0 +1,84 @@
...
+                n = int(args)
+            except ValueError:
+                return "Argument must be a number!"
+
+        karmees = karmees[:n]
+
+        suffixes = ['', 'st', 'nd', 'rd', 'th', 'th']
+
+        for pos, (k, v) in enumerate(karmees, start=1):
+            output += '{0}{1} {2} with {3}\n'.format(pos, suffixes[pos], v, k)
```
Comment (mrshu) [inline karma.py]: @Adman this won't fly. If I am reading it correctly, calling `!top_karma 7` will make this line trigger a `ValueError`.

File: karma.py [hunk 3]
```diff
@@ -0,0 +1,84 @@
...
+class Karma(BotPlugin):
+    def activate(self):
+        super(Karma, self).activate()
+        try:
+            if type(self['karma']) is not dict:
+                self['karma'] = {}
+        except KeyError:
+            self['karma'] = {}
+
+    def update_karma(self, msg, match, increment):
```
Comment (mrshu) [inline karma.py:17]: @Adman nicely done.

File: karma.py [hunk 4]
```diff
@@ -0,0 +1,84 @@
...
+        nick = msg.frm.nick
+        karmas = self['karma']
+
+        if name == nick:
+            return 'Not in this universe, maggot!'
+        else:
+            if name not in karmas:
+                karmas[name] = 0
+
+            if increment:
```
Comment (mrshu) [inline karma.py:29]: @Adman I could see how we could improve this a bit by doing something like

``` python
delta = len(update)
sign = -1 if increment else 1
karmas[name] += sign*delta
```

but it's probably not necessary as this is quite readable.

**Comment by mrshu** [comment]
@Adman Nicely done! I left a few comments, they are mostly nitpicks.

The only serious question I have is this: how about some [tests](http://errbot.io/en/latest/user_guide/plugin_development/testing.html?highlight=tests)?

**Comment by Adman** [comment]
@mrshu Done. In the future if I find out how to force the function not to process output with markdown formatting, I will rewrite ordinal number suffix with simple dot.

Tests are always advantage which would speed up the review process, so I agree with adding them to the repository.

**Comment by mrshu** [comment]
> Tests are always advantage which would speed up the review process, so I agree with adding them to the repository.

When can we expect them then ? =)

**Comment by Adman** [comment]
@mrshu I expect them to be added during this weekend :P

**Comment by mrshu** [comment]
@Adman all right, I do not want to delay the release of this for any longer.

LGTM

**Pull request merged**